posthog/temporal/data_modeling/run_workflow.py

```diff
@@ -0,0 +1,654 @@
+    the model is not marked as `selected`, then it doesn't need to be materialized, and we
+    can immediately put it back in the queue as `ModelStatus.COMPLETED`.
+
+    Args:
+        model: The model we are trying to run.
+        team_id: The ID of the team who owns this model.
+        queue: The execution queue where we will report back results.
+    """
+    try:
+        if model.selected is True:
+            team = await database_sync_to_async(Team.objects.get)(id=team_id)
+            await materialize_model(model.label, team)
+    except Exception:
+        await queue.put(QueueMessage(status=ModelStatus.FAILED, label=model.label))
+    else:
+        await queue.put(QueueMessage(status=ModelStatus.COMPLETED, label=model.label))
+    finally:
+        queue.task_done()
+
+
+async def materialize_model(model_label: str, team: Team) -> tuple[str, DeltaTable]:
+    """Materialize a given model by running its query in a dlt pipeline.
+
+    Arguments:
+        model_label: A label representing the ID or the name of the model to materialize.
+            If it's a valid UUID, then we will assume it's the ID, otherwise we'll assume
+            it is the model's name.
+        team: The team the model belongs to.
+    """
+    filter_params: dict[str, str | uuid.UUID] = {}
+    try:
+        model_id = uuid.UUID(model_label)
+        filter_params["id"] = model_id
+    except ValueError:
+        model_name = model_label
+        filter_params["name"] = model_name
+
+    saved_query = await database_sync_to_async(DataWarehouseSavedQuery.objects.filter(team=team, **filter_params).get)()
+
+    query_columns = saved_query.columns
+    if not query_columns:
+        query_columns = await database_sync_to_async(saved_query.get_columns)()
+
+    table_columns = {}
+    for column_name, column_info in query_columns.items():
+        clickhouse_type = column_info["clickhouse"]
+        nullable = False
+
+        if nullable_match := re.match(NullablePattern, clickhouse_type):
+            clickhouse_type = nullable_match.group(1)
+            nullable = True
+
+        clickhouse_type = re.sub(r"\(.+\)+", "", clickhouse_type)
+
+        table_columns[column_name] = {
+            "data_type": CLICKHOUSE_DLT_MAPPING[clickhouse_type],
+            "nullable": nullable,
+        }
+
+    hogql_query = saved_query.query["query"]
+
+    destination = get_dlt_destination()
+    pipeline = dlt.pipeline(
+        pipeline_name=f"materialize_model_{model_label}",
+        destination=destination,
+        dataset_name=f"team_{team.pk}_model_{model_label}",
+    )
+    _ = await asyncio.to_thread(pipeline.run, hogql_table(hogql_query, team, saved_query.name, table_columns))
+
+    tables = get_delta_tables(pipeline)
+    key, delta_table = tables.popitem()
+    return (key, delta_table)
+
+
+@dlt.source(max_table_nesting=0)
+def hogql_table(query: str, team: Team, table_name: str, table_columns):
```

Comment:
Very much a nit, but can we type `table_columns`?

Reply:
Quite annoying that dlt doesn't have flexible typing here though. I need to dig really deep to find those cryptic `TAnySchemaColumns` types.

Reply:
Ok, done in latest commit. That was very annoying and adds pretty much nothing. Dlt really needs to work on their duck typing and exposing all types in a top level `typing` to avoid having to dig through imports.